x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -34,35 +40,61 @@ pub fn get_balance(account: Address) -> u64 {
     borsh::from_slice(&bytes).expect("failed to deserialize the balance")
 }
 
-pub struct State<'a, K: Key> {
-    cache: &'a RefCell<HashMap<K, Option<Vec<u8>>>>,
+pub struct State<'a> {
+    cache: &'a RefCell<HashMap<CacheKey, Option<CacheValue>>>,
 }
 
 /// Key trait for program state keys
 /// # Safety
 /// This trait should only be implemented using the [`state_keys`](crate::state_keys) macro.
 pub unsafe trait Key: Copy + PartialEq + Eq + Hash + BorshSerialize {}
 
-impl<'a, K: Key> Drop for State<'a, K> {
-    fn drop(&mut self) {
-        if !self.cache.borrow().is_empty() {
-            // force flush
-            self.flush();
-        }
+#[derive(Clone, Copy)]
+#[repr(C, packed)]
+pub(crate) struct PrefixedKey<K: NoUninit> {
+    prefix: u8,
+    key: K,
+}
+
+impl<K: NoUninit> AsRef<[u8]> for PrefixedKey<K> {
+    fn as_ref(&self) -> &[u8] {
+        bytemuck::bytes_of(self)
+    }
+}
+
+// TODO: deal wiht padding?
+unsafe impl<K: NoUninit> NoUninit for PrefixedKey<K> {}
+
+// TODO: use bytemuck::must_cast (behind feature flag)
+pub unsafe trait Schema: NoUninit {
+    type Value: BorshSerialize + BorshDeserialize;
+
+    fn prefix() -> u8;
+
+    fn get(self, context: &mut Context) -> Result<Option<Self::Value>, Error> {
+        let key = to_key(self);
+        context.get_with_raw_key(key.as_ref())
```

Comment:
Is it safer not to expose `get_with_raw_key`? Or should this be an unsafe method?

Reply:
Deal with this later